<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>GeneScout: De Novo Small ORF Discovery • GeneScout</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GeneScout: De Novo Small ORF Discovery">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GeneScout</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/GeneScout.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/GeneScout_on_APOE-and-chr19.html">GeneScout on APOE and chr19</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/danymukesha/GeneScout/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GeneScout: De Novo Small ORF Discovery</h1>
                        <h4 data-toc-skip class="author">Dany
Mukesha</h4>
            
            <h4 data-toc-skip class="date">2026-01-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/danymukesha/GeneScout/blob/main/vignettes/GeneScout.Rmd" class="external-link"><code>vignettes/GeneScout.Rmd</code></a></small>
      <div class="d-none name"><code>GeneScout.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>A toolkit to identify hidden genes in non-coding DNA using
      statistical information theory. It implements sliding window
      entropy scanning, Shannon entropy calculation, and
      Kullback-Leibler divergence to discover potential small open
      reading frames (ORFs) in genomic sequences without requiring prior
      annotation.</p>
    </div>
    
<p><strong>The problem:</strong> Even in well-studied genomes, large
stretches of “junk” DNA actually contain small, hidden genes. These can
be identified by looking for patterns in codon usage using entropy
analysis.</p>
<ul>
<li>In non-coding DNA, codons appear randomly (high entropy)</li>
<li>In real genes, organisms prefer specific codons (low entropy)</li>
</ul>
<p><code>GeneScout</code> provides tools to calculate codon usage
profiles from known genes, scan large genomic regions with a sliding
window, identify regions with low entropy (potential coding regions),
and find candidate ORFs within those regions</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install development version</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"danymukesha/GeneScout"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="section level3">
<h3 id="basic-enthropy-calculation">Basic enthropy calculation<a class="anchor" aria-label="anchor" href="#basic-enthropy-calculation"></a>
</h3>
<p>First, calculate codon frequencies and Shannon entropy for a
sequence:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example sequence (biased codon usage)</span></span>
<span><span class="va">sequence1</span> <span class="op">&lt;-</span> <span class="st">"ATGATGATGATGATGATGTTATTATTATTATTATTATTACGCCGCCGCCGCCGCC"</span></span>
<span></span>
<span><span class="va">freqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_codon_frequencies.html">calculate_codon_frequencies</a></span><span class="op">(</span><span class="va">sequence1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">freqs</span><span class="op">[</span><span class="va">freqs</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       TTA       CGC       ATG </span></span>
<span><span class="co">## 0.3888889 0.2777778 0.3333333</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">entropy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_shannon_entropy.html">calculate_shannon_entropy</a></span><span class="op">(</span><span class="va">freqs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Shannon Entropy:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">entropy</span>, <span class="fl">3</span><span class="op">)</span>, <span class="st">"bits"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Shannon Entropy: 1.572 bits"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="create-a-reference-profile">Create a reference profile<a class="anchor" aria-label="anchor" href="#create-a-reference-profile"></a>
</h3>
<p>For de novo discovery, create a reference profile from known
genes:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Known gene sequences (would normally be from annotated genes)</span></span>
<span><span class="va">known_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"ATGATGATGATGCCGCCGCCGCCTTATTATTATTAG"</span>,</span>
<span>  <span class="st">"ATGCCGCCGCCGCCATTATTATTAATGTTTTTTTTT"</span>,</span>
<span>  <span class="st">"ATGTTTTTTTTTTTACGCCGCCGCCGCCGAGTAA"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create reference codon usage profile</span></span>
<span><span class="va">ref_profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_reference_profile.html">create_reference_profile</a></span><span class="op">(</span><span class="va">known_genes</span>, method <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ref_profile</span><span class="op">[</span><span class="va">ref_profile</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        CCA        CGA        GTA        TTA        CGC        TAG </span></span>
<span><span class="co">## 0.02777778 0.03030303 0.03030303 0.11363636 0.12121212 0.02777778</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="sliding-window-scan">Sliding window scan<a class="anchor" aria-label="anchor" href="#sliding-window-scan"></a>
</h3>
<p>Perform a genome-wide scan to find regions matching the reference
profile:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a longer test sequence with embedded "gene-like" regions</span></span>
<span><span class="va">test_sequence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"ACGTACGTACGT"</span>, <span class="fl">30</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,  <span class="co"># Random region</span></span>
<span>  <span class="st">"ATGATGATGATGCCGCCGCCGCCTTATTATTATTA"</span>,       <span class="co"># Embedded ORF</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"TGACTGACTGA"</span>, <span class="fl">30</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span>, <span class="co"># Random region</span></span>
<span>  <span class="st">"ATGCCGCCGCCGCCATTATTATTAATGTTTTTTTTTTAG"</span>,   <span class="co"># Another ORF</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"GATCGATCGAT"</span>, <span class="fl">30</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="co"># Random region</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Scan with sliding window</span></span>
<span><span class="va">scan_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sliding_window_scan.html">sliding_window_scan</a></span><span class="op">(</span></span>
<span>  <span class="va">test_sequence</span>,</span>
<span>  window_size <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  step_size <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  reference_profile <span class="op">=</span> <span class="va">ref_profile</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">scan_result</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GeneScout Sliding Window Scan Results</span></span>
<span><span class="co">## =======================================</span></span>
<span><span class="co">## Number of windows: 32 </span></span>
<span><span class="co">## Sequence range: 1 - 1080 bp</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Entropy Statistics:</span></span>
<span><span class="co">##   Mean Shannon Entropy: 2.665 bits</span></span>
<span><span class="co">##   Std. Dev. Shannon Entropy: 0.598 bits</span></span>
<span><span class="co">##   Min Shannon Entropy: 1.999 bits</span></span>
<span><span class="co">##   Max Shannon Entropy: 3.896 bits</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## KL Divergence Statistics:</span></span>
<span><span class="co">##   Mean KL Divergence: 23.452 </span></span>
<span><span class="co">##   Std. Dev. KL Divergence: 3.108</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="detect-entropy-peaks">Detect Entropy Peaks<a class="anchor" aria-label="anchor" href="#detect-entropy-peaks"></a>
</h3>
<p>Identify regions with unusually low entropy:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Detect peaks in entropy profile</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/entropy_peak_detection.html">entropy_peak_detection</a></span><span class="op">(</span></span>
<span>  <span class="va">scan_result</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"shannon_entropy"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"quantile"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_peak_width <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   peak_id start_window end_window num_windows start_bp end_bp metric_value_mean</span></span>
<span><span class="co">## 1       1            1          8           8        1    360          1.998846</span></span>
<span><span class="co">##   metric_value_min</span></span>
<span><span class="co">## 1         1.998846</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="find-candidate-orfs">Find candidate ORFs<a class="anchor" aria-label="anchor" href="#find-candidate-orfs"></a>
</h3>
<p>Search for actual ORFs within the low-entropy peaks:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find candidate ORFs</span></span>
<span><span class="va">candidates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_candidate_orfs.html">find_candidate_orfs</a></span><span class="op">(</span></span>
<span>  <span class="va">test_sequence</span>,</span>
<span>  <span class="va">scan_result</span>,</span>
<span>  <span class="va">peaks</span>,</span>
<span>  min_orf_length <span class="op">=</span> <span class="fl">60</span>,</span>
<span>  start_codons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ATG"</span><span class="op">)</span>,</span>
<span>  stop_codons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TAA"</span>, <span class="st">"TAG"</span>, <span class="st">"TGA"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">candidates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 0 × 0</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualize-results">Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h3>
<p>Create plots to visualize entropy profiles and candidate ORFs:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot entropy profile</span></span>
<span><span class="fu"><a href="../reference/plot_entropy_profile.html">plot_entropy_profile</a></span><span class="op">(</span><span class="va">scan_result</span>, </span>
<span>                     metric <span class="op">=</span> <span class="st">"shannon_entropy"</span>,</span>
<span>                     highlight_threshold <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneScout_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" alt="" width="960"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot with peaks highlighted</span></span>
<span><span class="fu"><a href="../reference/plot_entropy_profile.html">plot_entropy_profile</a></span><span class="op">(</span><span class="va">scan_result</span>, </span>
<span>                     peaks <span class="op">=</span> <span class="va">peaks</span>,</span>
<span>                     metric <span class="op">=</span> <span class="st">"shannon_entropy"</span>,</span>
<span>                     show_peaks <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneScout_files/figure-html/unnamed-chunk-8-1.png" class="r-plt" alt="" width="960"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot candidate ORFs</span></span>
<span><span class="fu"><a href="../reference/plot_candidate_orfs.html">plot_candidate_orfs</a></span><span class="op">(</span><span class="va">scan_result</span>, <span class="va">candidates</span>, <span class="va">peaks</span>, metric <span class="op">=</span> <span class="st">"shannon_entropy"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneScout_files/figure-html/unnamed-chunk-9-1.png" class="r-plt" alt="" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="advanced-usage">Advanced usage<a class="anchor" aria-label="anchor" href="#advanced-usage"></a>
</h2>
<div class="section level3">
<h3 id="comparative-analysis">Comparative analysis<a class="anchor" aria-label="anchor" href="#comparative-analysis"></a>
</h3>
<p>Compare entropy profiles between different regions:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create two different sequences</span></span>
<span><span class="co"># APOE coding region (simulated longer fragment for demo, ~4 kb)</span></span>
<span><span class="va">seq_region1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="st">"ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCAGGAGAGCTGCGGGAGGAGCGGAGGCTG"</span>,</span>
<span>    <span class="st">"GAGGCCGCGGAGGAGCTGGCGGAGGCGGAGGAGGCGGAGGAGCCGCGGAGGAGGAGGAGGAG"</span>,</span>
<span>    <span class="st">"GCCGCGGAGGAGGCGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span>,</span>
<span>    <span class="st">"GAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span>,</span>
<span>    <span class="st">"GAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span>,</span>
<span>    <span class="st">"GCCGCGGAGGAGGCGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span>,</span>
<span>    <span class="st">"GAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span>,</span>
<span>    <span class="st">"GAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span>,</span>
<span>    <span class="st">"GAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGAGGA"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Random-like region from chr19 (simulated, ~4 kb)</span></span>
<span><span class="va">seq_region2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="st">"GCGTACGTAGCTAGCGTAGCTACGTAGCTAGCTGACGATCAGTGCTAGCTAGCTAGCGTAC"</span>,</span>
<span>    <span class="st">"GATCGATCGATCGTAGCTAGCTAGCTAGCGTACGATCGATCGTAGCTAGCTAGCTGACGAT"</span>,</span>
<span>    <span class="st">"CGTAGCTAGCTGACGATCGTACGATGCTAGCTAGCTGATCGTAGCTAGCTAGCTAGCGTAC"</span>,</span>
<span>    <span class="st">"GATCGATCGTAGCTAGCTAGCTGACGATCGTAGCTAGCTGACGATCGTACGATGCTAGCTA"</span>,</span>
<span>    <span class="st">"TACGATCGATCGATCGTAGCTAGCTAGCTAGCGTACGATCGATCGTAGCTAGCTAGCTGAC"</span>,</span>
<span>    <span class="st">"GATCGTAGCTAGCTGACGATCGTACGATGCTAGCTAGCTGATCGTAGCTAGCTAGCTAGCG"</span>,</span>
<span>    <span class="st">"TACGATCGATCGTAGCTAGCTAGCTGACGATCGTAGCTAGCTGACGATCGTACGATGCTAG"</span>,</span>
<span>    <span class="st">"CTATACGATCGATCGATCGTAGCTAGCTAGCTAGCGTACGATCGATCGTAGCTAGCTAGCT"</span>,</span>
<span>    <span class="st">"GACGATCGTAGCTAGCTGACGATCGTACGATGCTAGCTAGCTGATCGTAGCTAGCTAGCTA"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scan both regions</span></span>
<span><span class="va">result1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sliding_window_scan.html">sliding_window_scan</a></span><span class="op">(</span><span class="va">seq_region1</span>, window_size <span class="op">=</span> <span class="fl">100</span>, step_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">result2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sliding_window_scan.html">sliding_window_scan</a></span><span class="op">(</span><span class="va">seq_region2</span>, window_size <span class="op">=</span> <span class="fl">100</span>, step_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare profiles</span></span>
<span><span class="fu"><a href="../reference/compare_entropy_profiles.html">compare_entropy_profiles</a></span><span class="op">(</span><span class="va">result1</span>, <span class="va">result2</span>, </span>
<span>                         labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene-like region"</span>, <span class="st">"Random region"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneScout_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" alt="" width="700"></p>
</div>
<div class="section level3">
<h3 id="working-with-genomes">Working with genomes<a class="anchor" aria-label="anchor" href="#working-with-genomes"></a>
</h3>
<p>For large genomes, process chromosome by chromosome:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read genome FASTA file (https://parasite.wormbase.org/index.html)</span></span>
<span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fasta.html">read_fasta</a></span><span class="op">(</span><span class="st">"../data/acanthocheilonema_viteae.PRJEB1697.WBPS19.genomic.fa"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract known genes from GTF annotation file</span></span>
<span><span class="va">known_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_known_genes.html">extract_known_genes</a></span><span class="op">(</span></span>
<span>  gtf_file <span class="op">=</span> <span class="st">"../data/acanthocheilonema_viteae.PRJEB1697.WBPS19.canonical_geneset.gtf"</span>,</span>
<span>  genome_fasta <span class="op">=</span> <span class="st">"../data/acanthocheilonema_viteae.PRJEB1697.WBPS19.genomic.fa"</span>,</span>
<span>  feature_type <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  min_length <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  max_genes <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create reference profile from extracted genes</span></span>
<span><span class="va">ref_profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_reference_profile.html">create_reference_profile</a></span><span class="op">(</span><span class="va">known_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Process one chromosome</span></span>
<span><span class="va">chr_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">chr_sequence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">genome</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scan chromosome</span></span>
<span><span class="va">chr_scan</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sliding_window_scan.html">sliding_window_scan</a></span><span class="op">(</span></span>
<span>  <span class="va">chr_sequence</span>,</span>
<span>  window_size <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  step_size <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  reference_profile <span class="op">=</span> <span class="va">ref_profile</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Detect peaks and find ORFs</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/entropy_peak_detection.html">entropy_peak_detection</a></span><span class="op">(</span><span class="va">chr_scan</span>, threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">candidates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_candidate_orfs.html">find_candidate_orfs</a></span><span class="op">(</span><span class="va">chr_sequence</span>, <span class="va">chr_scan</span>, <span class="va">peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">candidates</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"candidates_"</span>, <span class="va">chr_name</span>, <span class="st">".csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="theory-behind-genescout">Theory behind GeneScout<a class="anchor" aria-label="anchor" href="#theory-behind-genescout"></a>
</h2>
<div class="section level3">
<h3 id="shannon-enthropy">Shannon enthropy<a class="anchor" aria-label="anchor" href="#shannon-enthropy"></a>
</h3>
<p>Shannon entropy measures the “surprise” or randomness of data:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mo>=</mo><mo>−</mo><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mn>64</mn></munderover><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><msub><mo>log</mo><mn>2</mn></msub><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">H = -\sum_{i=1}^{64} P(i) \log_2 P(i)</annotation></semantics></math></p>
<p>Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(i)</annotation></semantics></math>
is the frequency of codon
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.</p>
<ul>
<li>
<strong>High entropy</strong> (~6 bits): Random codon usage
(non-coding DNA)</li>
<li>
<strong>Low entropy</strong> (~3-4 bits): Biased codon usage
(potential gene)</li>
</ul>
</div>
<div class="section level3">
<h3 id="kullback-leibler-divergance">Kullback-Leibler divergance<a class="anchor" aria-label="anchor" href="#kullback-leibler-divergance"></a>
</h3>
<p>KL divergence measures how one distribution differs from another:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>D</mi><mrow><mi>K</mi><mi>L</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>P</mi><mo stretchy="false" form="postfix">∥</mo><mi>Q</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><munder><mo>∑</mo><mi>i</mi></munder><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><msub><mo>log</mo><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><mi>Q</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">D_{KL}(P \| Q) = \sum_{i} P(i) \log_2 \left( \frac{P(i)}{Q(i)} \right)</annotation></semantics></math></p>
<p>Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(i)</annotation></semantics></math>
is the observed frequency and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Q</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Q(i)</annotation></semantics></math>
is the reference frequency.</p>
<ul>
<li>
<strong>High KL</strong>: Window differs from organism’s typical
codon usage</li>
<li>
<strong>Low KL</strong>: Window matches organism’s typical codon
usage</li>
</ul>
</div>
<div class="section level3">
<h3 id="effective-number-of-codons-enc">Effective Number of Codons (ENC)<a class="anchor" aria-label="anchor" href="#effective-number-of-codons-enc"></a>
</h3>
<p>ENC is another measure of codon usage bias:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate ENC for different sequences</span></span>
<span><span class="va">seq_uniform</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ATG"</span>, <span class="st">"TTT"</span>, <span class="st">"CCC"</span>, <span class="st">"GGG"</span>, <span class="st">"AAA"</span>, <span class="st">"TTT"</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span>, </span>
<span>                     collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span><span class="va">seq_biased</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"ATG"</span>, <span class="fl">60</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="va">freqs_uniform</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_codon_frequencies.html">calculate_codon_frequencies</a></span><span class="op">(</span><span class="va">seq_uniform</span><span class="op">)</span></span>
<span><span class="va">freqs_biased</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_codon_frequencies.html">calculate_codon_frequencies</a></span><span class="op">(</span><span class="va">seq_biased</span><span class="op">)</span></span>
<span></span>
<span><span class="va">enc_uniform</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_enc.html">calculate_enc</a></span><span class="op">(</span><span class="va">freqs_uniform</span><span class="op">)</span></span>
<span><span class="va">enc_biased</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_enc.html">calculate_enc</a></span><span class="op">(</span><span class="va">freqs_biased</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Uniform sequence ENC:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">enc_uniform</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Uniform sequence ENC: 61"</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Biased sequence ENC:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">enc_biased</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Biased sequence ENC: 61"</span></span></code></pre>
<p>ENC ranges from 20 (extreme bias) to 61 (no bias).</p>
</div>
</div>
<div class="section level2">
<h2 id="parameter-tuning">Parameter tuning<a class="anchor" aria-label="anchor" href="#parameter-tuning"></a>
</h2>
<div class="section level3">
<h3 id="window-size">Window size<a class="anchor" aria-label="anchor" href="#window-size"></a>
</h3>
<p>The choice of window size affects sensitivity:</p>
<ul>
<li>
<strong>Small windows</strong> (50-100 bp): Higher sensitivity, more
noise</li>
<li>
<strong>Large windows</strong> (300-500 bp): More stable, may miss
small ORFs</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare different window sizes</span></span>
<span><span class="va">windows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">150</span>, <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">windows</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ws</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/sliding_window_scan.html">sliding_window_scan</a></span><span class="op">(</span><span class="va">test_sequence</span>, window_size <span class="op">=</span> <span class="va">ws</span>, step_size <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-size">Step size<a class="anchor" aria-label="anchor" href="#step-size"></a>
</h3>
<p>Step size determines resolution:</p>
<ul>
<li>
<strong>Small step</strong> (10-20 bp): Fine-grained scan,
slower</li>
<li>
<strong>Large step</strong> (30-50 bp): Faster scan, may miss
features</li>
</ul>
</div>
<div class="section level3">
<h3 id="threshold-selection">Threshold selection<a class="anchor" aria-label="anchor" href="#threshold-selection"></a>
</h3>
<p>Use different methods to set entropy thresholds:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Quantile-based threshold</span></span>
<span><span class="va">peaks_quantile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/entropy_peak_detection.html">entropy_peak_detection</a></span><span class="op">(</span></span>
<span>  <span class="va">scan_result</span>, </span>
<span>  method <span class="op">=</span> <span class="st">"quantile"</span>, </span>
<span>  threshold <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Standard deviation-based threshold</span></span>
<span><span class="va">peaks_sd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/entropy_peak_detection.html">entropy_peak_detection</a></span><span class="op">(</span></span>
<span>  <span class="va">scan_result</span>, </span>
<span>  method <span class="op">=</span> <span class="st">"sd"</span>, </span>
<span>  threshold <span class="op">=</span> <span class="fl">1.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="integration-with-rna-seq">Integration with RNA-Seq<a class="anchor" aria-label="anchor" href="#integration-with-rna-seq"></a>
</h2>
<p>For validation, integrate with RNA-seq expression data:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load RNA-seq coverage data</span></span>
<span><span class="va">coverage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"rnaseq_coverage.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add expression information to candidates</span></span>
<span><span class="va">candidates_with_expr</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">candidates</span>,</span>
<span>  mean_coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">candidates</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">coverage</span><span class="op">[</span><span class="va">candidates</span><span class="op">$</span><span class="va">start</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">:</span><span class="va">candidates</span><span class="op">$</span><span class="va">end</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter by expression</span></span>
<span><span class="va">expressed_candidates</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>  <span class="va">candidates_with_expr</span>,</span>
<span>  <span class="va">mean_coverage</span> <span class="op">&gt;</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="performance-consideration">Performance consideration<a class="anchor" aria-label="anchor" href="#performance-consideration"></a>
</h2>
<p>For large genomes:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Process by chromosome</strong>: Avoid loading entire genome
into memory</li>
<li>
<strong>Use parallel processing</strong>: Scan different chromosomes
in parallel</li>
<li>
<strong>Optimize parameters</strong>: Larger step sizes reduce
computation</li>
<li>
<strong>Filter early</strong>: Exclude low-complexity regions before
scanning</li>
</ol>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Parallel scanning example</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RevolutionAnalytics/foreach" class="external-link">foreach</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RevolutionAnalytics/doparallel" class="external-link">doParallel</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">registerDoParallel</span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">foreach</span><span class="op">(</span>chr <span class="op">=</span> <span class="va">chromosome_names</span><span class="op">)</span> <span class="op">%dopar%</span> <span class="op">{</span></span>
<span>  <span class="va">chr_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">genome</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/sliding_window_scan.html">sliding_window_scan</a></span><span class="op">(</span><span class="va">chr_seq</span>, window_size <span class="op">=</span> <span class="fl">300</span>, step_size <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Wright, F. (1990). The ‘effective number of codons’ used in a gene.
<em>Gene</em>, 87(1), 23-29.</li>
<li>Sharp, P. M., &amp; Li, W. H. (1987). The codon Adaptation Index-a
measure of directional synonymous codon usage bias, and its potential
applications. <em>Nucleic Acids Research</em>, 15(3), 1281-1295.</li>
</ul>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] GeneScout_0.1.0  BiocStyle_2.38.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] SummarizedExperiment_1.40.0 gtable_0.3.6               </span></span>
<span><span class="co">##  [3] rjson_0.2.23                xfun_0.56                  </span></span>
<span><span class="co">##  [5] bslib_0.9.0                 ggplot2_4.0.1              </span></span>
<span><span class="co">##  [7] lattice_0.22-7              Biobase_2.70.0             </span></span>
<span><span class="co">##  [9] vctrs_0.7.0                 tools_4.5.2                </span></span>
<span><span class="co">## [11] bitops_1.0-9                generics_0.1.4             </span></span>
<span><span class="co">## [13] stats4_4.5.2                curl_7.0.0                 </span></span>
<span><span class="co">## [15] parallel_4.5.2              tibble_3.3.1               </span></span>
<span><span class="co">## [17] pkgconfig_2.0.3             Matrix_1.7-4               </span></span>
<span><span class="co">## [19] RColorBrewer_1.1-3          cigarillo_1.0.0            </span></span>
<span><span class="co">## [21] S7_0.2.1                    desc_1.4.3                 </span></span>
<span><span class="co">## [23] S4Vectors_0.48.0            lifecycle_1.0.5            </span></span>
<span><span class="co">## [25] compiler_4.5.2              farver_2.1.2               </span></span>
<span><span class="co">## [27] stringr_1.6.0               Rsamtools_2.26.0           </span></span>
<span><span class="co">## [29] textshaping_1.0.4           Biostrings_2.78.0          </span></span>
<span><span class="co">## [31] Seqinfo_1.0.0               codetools_0.2-20           </span></span>
<span><span class="co">## [33] htmltools_0.5.9             sass_0.4.10                </span></span>
<span><span class="co">## [35] RCurl_1.98-1.17             yaml_2.3.12                </span></span>
<span><span class="co">## [37] pillar_1.11.1               pkgdown_2.2.0              </span></span>
<span><span class="co">## [39] crayon_1.5.3                jquerylib_0.1.4            </span></span>
<span><span class="co">## [41] tidyr_1.3.2                 BiocParallel_1.44.0        </span></span>
<span><span class="co">## [43] DelayedArray_0.36.0         cachem_1.1.0               </span></span>
<span><span class="co">## [45] abind_1.4-8                 tidyselect_1.2.1           </span></span>
<span><span class="co">## [47] digest_0.6.39               stringi_1.8.7              </span></span>
<span><span class="co">## [49] dplyr_1.1.4                 purrr_1.2.1                </span></span>
<span><span class="co">## [51] restfulr_0.0.16             bookdown_0.46              </span></span>
<span><span class="co">## [53] labeling_0.4.3              fastmap_1.2.0              </span></span>
<span><span class="co">## [55] grid_4.5.2                  SparseArray_1.10.8         </span></span>
<span><span class="co">## [57] cli_3.6.5                   magrittr_2.0.4             </span></span>
<span><span class="co">## [59] S4Arrays_1.10.1             XML_3.99-0.20              </span></span>
<span><span class="co">## [61] withr_3.0.2                 scales_1.4.0               </span></span>
<span><span class="co">## [63] rmarkdown_2.30              XVector_0.50.0             </span></span>
<span><span class="co">## [65] httr_1.4.7                  matrixStats_1.5.0          </span></span>
<span><span class="co">## [67] ragg_1.5.0                  evaluate_1.0.5             </span></span>
<span><span class="co">## [69] knitr_1.51                  GenomicRanges_1.62.1       </span></span>
<span><span class="co">## [71] IRanges_2.44.0              BiocIO_1.20.0              </span></span>
<span><span class="co">## [73] rtracklayer_1.70.1          rlang_1.1.7                </span></span>
<span><span class="co">## [75] glue_1.8.0                  BiocManager_1.30.27        </span></span>
<span><span class="co">## [77] BiocGenerics_0.56.0         jsonlite_2.0.0             </span></span>
<span><span class="co">## [79] R6_2.6.1                    MatrixGenerics_1.22.0      </span></span>
<span><span class="co">## [81] GenomicAlignments_1.46.0    systemfonts_1.3.1          </span></span>
<span><span class="co">## [83] fs_1.6.6</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dany Mukesha.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
